File: Sources/DimaSave/Models/AppPaths.swift
import Foundation

enum AppPaths {
    static let appName = "DimaSave"
    private static let embeddedRuntimeDirectoryName = "runtime"
    private static let resourceBundleName = "DimaSave_DimaSave.bundle"

    static var homeDirectory: URL {
        FileManager.default.homeDirectoryForCurrentUser
    }

    private static var preferredApplicationSupport: URL {
        homeDirectory
            .appendingPathComponent("Library", isDirectory: true)
            .appendingPathComponent("Application Support", isDirectory: true)
            .appendingPathComponent(appName, isDirectory: true)
    }

    private static var runtimeFallback: URL {
        URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
            .appendingPathComponent(".runtime", isDirectory: true)
            .appendingPathComponent(appName, isDirectory: true)
    }

    private static func canWrite(to directory: URL) -> Bool {
        let fileManager = FileManager.default
        var candidate = directory

        while candidate.path != "/" {
            var isDirectory: ObjCBool = false
            if fileManager.fileExists(atPath: candidate.path, isDirectory: &isDirectory) {
                return isDirectory.boolValue && fileManager.isWritableFile(atPath: candidate.path)
            }
            candidate.deleteLastPathComponent()
        }

        return false
    }

    static var applicationSupport: URL {
        if let override = ProcessInfo.processInfo.environment["DIMASAVE_APP_SUPPORT"] {
            return URL(fileURLWithPath: override, isDirectory: true)
        }

        return canWrite(to: preferredApplicationSupport.deletingLastPathComponent())
            ? preferredApplicationSupport
            : runtimeFallback
    }

    private static var installedWorkerRoot: URL? {
        let fileManager = FileManager.default
        let candidate = preferredApplicationSupport.appendingPathComponent("worker", isDirectory: true)
        let python = candidate
            .appendingPathComponent(".venv", isDirectory: true)
            .appendingPathComponent("bin", isDirectory: true)
            .appendingPathComponent("python3", isDirectory: false)

        return fileManager.fileExists(atPath: python.path) ? candidate : nil
    }

    static var workerRoot: URL {
        applicationSupport.appendingPathComponent("worker", isDirectory: true)
    }

    static var workerVenvRoot: URL {
        (installedWorkerRoot ?? workerRoot).appendingPathComponent(".venv", isDirectory: true)
    }

    static var workerPython: URL {
        if let bundled = bundledPythonExecutable {
            return bundled
        }

        return workerVenvRoot
            .appendingPathComponent("bin", isDirectory: true)
            .appendingPathComponent("python3", isDirectory: false)
    }

    static var browserProfile: URL {
        workerRoot.appendingPathComponent("browser-profile", isDirectory: true)
    }

    static var manifestsDirectory: URL {
        applicationSupport.appendingPathComponent("manifests", isDirectory: true)
    }

    static var playwrightBrowsers: URL {
        if let bundled = bundledPlaywrightBrowsers {
            return bundled
        }

        return (installedWorkerRoot ?? workerRoot).appendingPathComponent("ms-playwright", isDirectory: true)
    }

    static var bundledPythonHome: URL? {
        let path = bundledFrameworksRoot?
            .appendingPathComponent("Python.framework", isDirectory: true)
            .appendingPathComponent("Versions", isDirectory: true)
            .appendingPathComponent("3.13", isDirectory: true)

        return existingDirectory(at: path)
    }

    static var bundledPythonExecutable: URL? {
        let path = bundledPythonHome?
            .appendingPathComponent("bin", isDirectory: true)
            .appendingPathComponent("python3.13", isDirectory: false)

        return existingFile(at: path)
    }

    static var bundledSitePackages: URL? {
        let path = bundledRuntimeRoot?
            .appendingPathComponent("site-packages", isDirectory: true)

        return existingDirectory(at: path)
    }

    static var bundledPlaywrightBrowsers: URL? {
        let path = bundledRuntimeRoot?
            .appendingPathComponent("ms-playwright", isDirectory: true)

        return existingDirectory(at: path)
    }

    static var bundledFrameworksDirectory: URL? {
        bundledFrameworksRoot
    }

    static var hasEmbeddedRuntime: Bool {
        bundledPythonExecutable != nil && bundledSitePackages != nil && bundledPlaywrightBrowsers != nil
    }

    static var logsDirectory: URL {
        applicationSupport.appendingPathComponent("logs", isDirectory: true)
    }

    static var defaultDownloads: URL {
        let preferred = homeDirectory
            .appendingPathComponent("Downloads", isDirectory: true)
            .appendingPathComponent(appName, isDirectory: true)

        return canWrite(to: preferred.deletingLastPathComponent())
            ? preferred
            : applicationSupport.appendingPathComponent("Downloads", isDirectory: true)
    }

    static func ensureDirectories() throws {
        let fileManager = FileManager.default
        try fileManager.createDirectory(at: applicationSupport, withIntermediateDirectories: true)
        try fileManager.createDirectory(at: workerRoot, withIntermediateDirectories: true)
        try fileManager.createDirectory(at: browserProfile, withIntermediateDirectories: true)
        try fileManager.createDirectory(at: manifestsDirectory, withIntermediateDirectories: true)
        try fileManager.createDirectory(at: logsDirectory, withIntermediateDirectories: true)
        try fileManager.createDirectory(at: defaultDownloads, withIntermediateDirectories: true)

        if bundledPlaywrightBrowsers == nil {
            try fileManager.createDirectory(at: playwrightBrowsers, withIntermediateDirectories: true)
        }
    }

    private static var bundledFrameworksRoot: URL? {
        let candidates = [
            Bundle.main.privateFrameworksURL,
            Bundle.main.bundleURL
                .appendingPathComponent("Contents", isDirectory: true)
                .appendingPathComponent("Frameworks", isDirectory: true),
        ]

        return candidates.compactMap(existingDirectory(at:)).first
    }

    private static var bundledRuntimeRoot: URL? {
        let candidates = [
            Bundle.main.sharedSupportURL?
                .appendingPathComponent(embeddedRuntimeDirectoryName, isDirectory: true),
            Bundle.main.bundleURL
                .appendingPathComponent("Contents", isDirectory: true)
                .appendingPathComponent("SharedSupport", isDirectory: true)
                .appendingPathComponent(embeddedRuntimeDirectoryName, isDirectory: true),
            Bundle.main.resourceURL?
                .appendingPathComponent(resourceBundleName, isDirectory: true)
                .appendingPathComponent(embeddedRuntimeDirectoryName, isDirectory: true),
        ]

        return candidates.compactMap(existingDirectory(at:)).first
    }

    private static func existingDirectory(at url: URL?) -> URL? {
        guard let url else { return nil }
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory), isDirectory.boolValue else {
            return nil
        }
        return url
    }

    private static func existingFile(at url: URL?) -> URL? {
        guard let url else { return nil }
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory), !isDirectory.boolValue else {
            return nil
        }
        return url
    }
}

File: windows_app/dimasave_windows/app_paths.py
from __future__ import annotations

import os
import sys
from pathlib import Path


class AppPaths:
    app_name = "DimaSave"

    @staticmethod
    def project_root() -> Path:
        return Path(__file__).resolve().parents[2]

    @staticmethod
    def resource_root() -> Path:
        if getattr(sys, "frozen", False):
            return Path(getattr(sys, "_MEIPASS", Path(sys.executable).resolve().parent))
        return AppPaths.project_root()

    @staticmethod
    def application_support() -> Path:
        override = os.environ.get("DIMASAVE_APP_SUPPORT")
        if override:
            return Path(override).expanduser()

        root = os.environ.get("LOCALAPPDATA") or os.environ.get("APPDATA")
        if root:
            return Path(root) / AppPaths.app_name
        return Path.home() / "AppData" / "Local" / AppPaths.app_name

    @staticmethod
    def worker_root() -> Path:
        return AppPaths.application_support() / "worker"

    @staticmethod
    def worker_python() -> Path:
        return AppPaths.worker_root() / ".venv" / "Scripts" / "python.exe"

    @staticmethod
    def node_executable() -> Path:
        candidates = [
            AppPaths.resource_root() / "runtime" / "node" / "node.exe",
            AppPaths.resource_root() / "runtime" / "node" / "bin" / "node.exe",
        ]
        for candidate in candidates:
            if candidate.exists():
                return candidate
        raise FileNotFoundError("Не удалось найти встроенный runtime/node/node.exe")

    @staticmethod
    def node_worker_script() -> Path:
        candidates = [
            AppPaths.resource_root() / "node_worker" / "bridge.mjs",
            AppPaths.project_root() / "node_worker" / "bridge.mjs",
        ]
        for candidate in candidates:
            if candidate.exists():
                return candidate
        raise FileNotFoundError("Не удалось найти node_worker/bridge.mjs")

    @staticmethod
    def browser_profile() -> Path:
        return AppPaths.worker_root() / "browser-profile"

    @staticmethod
    def manifests_directory() -> Path:
        return AppPaths.application_support() / "manifests"

    @staticmethod
    def logs_directory() -> Path:
        return AppPaths.application_support() / "logs"

    @staticmethod
    def playwright_browsers() -> Path:
        bundled = AppPaths.resource_root() / "runtime" / "ms-playwright"
        if bundled.exists():
            return bundled
        return AppPaths.worker_root() / "ms-playwright"

    @staticmethod
    def has_embedded_runtime() -> bool:
        try:
            AppPaths.node_executable()
            AppPaths.node_worker_script()
            return AppPaths.playwright_browsers().exists()
        except FileNotFoundError:
            return False

    @staticmethod
    def default_downloads() -> Path:
        user_profile = os.environ.get("USERPROFILE")
        if user_profile:
            return Path(user_profile) / "Downloads" / AppPaths.app_name
        return AppPaths.application_support() / "Downloads"

    @staticmethod
    def worker_script() -> Path:
        candidates = [
            AppPaths.resource_root() / "worker" / "bridge.py",
            AppPaths.project_root() / "Sources" / "DimaSave" / "Resources" / "worker" / "bridge.py",
        ]
        for candidate in candidates:
            if candidate.exists():
                return candidate
        raise FileNotFoundError("Не удалось найти worker/bridge.py")

    @staticmethod
    def bootstrap_script() -> Path:
        candidates = [
            AppPaths.resource_root() / "windows_app" / "bootstrap_node_worker.ps1",
            AppPaths.project_root() / "windows_app" / "bootstrap_node_worker.ps1",
            AppPaths.resource_root() / "windows_app" / "bootstrap_worker.ps1",
            AppPaths.project_root() / "windows_app" / "bootstrap_worker.ps1",
        ]
        for candidate in candidates:
            if candidate.exists():
                return candidate
        raise FileNotFoundError("Не удалось найти bootstrap_worker.ps1")

    @staticmethod
    def ensure_directories() -> None:
        for path in [
            AppPaths.application_support(),
            AppPaths.worker_root(),
            AppPaths.browser_profile(),
            AppPaths.manifests_directory(),
            AppPaths.logs_directory(),
            AppPaths.playwright_browsers(),
            AppPaths.default_downloads(),
        ]:
            path.mkdir(parents=True, exist_ok=True)
